source/common/config/watch_map.cc

```diff
@@ -93,10 +93,36 @@ void WatchMap::onConfigUpdate(const Protobuf::RepeatedPtrField<ProtobufWkt::Any>
   }
 }
 
+// For responses to on-demand requests, replace the original watch for an alias
+// with one for the resource's name
+void WatchMap::convertAliasWatchToNameWatch(const envoy::api::v2::Resource& resource) {
+  absl::flat_hash_set<Watch*> watches_to_update;
+  for (const auto& alias : resource.aliases()) {
+    const auto interested_watches = watch_interest_.find(alias);
+    if (interested_watches != watch_interest_.end()) {
+      for (const auto& interested_watch : interested_watches->second) {
+        watches_to_update.insert(interested_watch);
+      }
+      break;
+    }
+  }
+  for (const auto& watch : watches_to_update) {
+    updateWatchInterest(watch, {resource.name()});
```

Comment:
Calling `updateWatchInterest(watch, {resource.name()})` will replace any existing resource names that watch was watching. Is your approach meant to assume that these alias resolutions will set a user's watch to care about just a single name? If so, you should carefully document that, so people who might expect otherwise don't get tripped up.

Although, I'm wondering, might it be better to add to rather than replace? If the idea is that the watch had been watching only aliases, which should all be replaced now that the canonical name is available, that makes sense... but that then relies on the server never doing anything with the alias again, since we'll no longer know how to handle that alias. Maybe that's fine; just seems like something that could be fragile.

Reply:
I played with the idea of leaving the alias around (and even updating watch interest to include all aliases), but that raises a few questions/issues:
 - why track aliases outside of request/response? Once resolved, the alias won't be used again. This doesn't cover aliases that couldn't be resolved, but that's a different scenario.
 - Aliases can potentially go away or new ones can appear (VHDS aliases are built from the domains list of a vhost, and nothing prevents from it being updated). Keeping watch interest in sync will be somewhat expensive, but also messy (every VHDS update, on-demand or not will have to reconcile them).

In the end I opted for keeping things simple, at least for the time being and overwrite aliases with corresponding resource names.

Reply:
I see, that makes sense, especially given "once resolved, the alias won't be used again". It should get some careful commenting/documenting to make clear what's going on and what's being assumed, but it does seem like the right way.